From d4b997bc051698a7e88cfe93a2db72d8d9607a3d Mon Sep 17 00:00:00 2001
From: Tommi Franttila <tommi.franttila@intel.com>
Date: Tue, 25 Nov 2014 13:32:04 +0200
Subject: [PATCH 2/8] atomisp2: Fix allocation of metadata list

If snapshots are taken during video the memory allocated by
vmap is leaked. This is because atomisp_reqbuf is called every
time snapshot is taken and new metadata buffers are incorrectly
allocated. To prevent memory leaking it should be also checked if
metadata_ready and metadata_in_css lists are empty or not and
perform allocation only if necessary. Fix it.

Change-Id: I7ba68e1fbcc7cb628636e9af6c32f5602a64d0a1
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-7959
Signed-off-by: Tommi Franttila <tommi.franttila@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/30114
Reviewed-by: Lassi I Vaatamoinen <lassi.i.vaatamoinen@intel.com>
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp_ioctl.c    |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
index 288fab1..31739d0 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
@@ -1025,7 +1025,9 @@ int atomisp_alloc_css_stat_bufs(struct atomisp_sub_device *asd,
 	}
 
 	for (i = 0; i < ATOMISP_METADATA_TYPE_NUM; i++) {
-		if (list_empty(&asd->metadata[i])) {
+		if (list_empty(&asd->metadata[i]) &&
+		    list_empty(&asd->metadata_ready[i]) &&
+		    list_empty(&asd->metadata_in_css[i])) {
 			count = ATOMISP_CSS_Q_DEPTH +
 				ATOMISP_METADATA_QUEUE_DEPTH_FOR_HAL;
 			dev_dbg(isp->dev, "allocating %d metadata buffers for type %d\n",
-- 
1.7.9.5

